pkg/http/delayed_writer.go

```diff
@@ -0,0 +1,67 @@
+package http
+
+import (
+	"net/http"
+)
+
+var (
+	_ http.Flusher        = (*DelayedWriter)(nil)
+	_ http.ResponseWriter = (*DelayedWriter)(nil)
+)
+
+// DelayedWriter is an implementation of http.ResponseWriter. It delays writing to the output until RealFlush is called.
+type DelayedWriter struct {
+	ResponseCode int
+
+	writer         http.ResponseWriter
+	headerRecorded bool
+
+	buffer        []byte
+	bytesBuffered int
+}
+
+// NewDelayedWriter creates an http.ResponseWriter that delays writing.
+func NewDelayedWriter(w http.ResponseWriter) *DelayedWriter {
+	return &DelayedWriter{
+		writer: w,
+	}
+}
+
+// Flush does nothing.
+func (dw *DelayedWriter) Flush() {}
+
+// RealFlush flushes the buffer (both header and body) to the client.
+func (dw *DelayedWriter) RealFlush() {
+	dw.writer.WriteHeader(dw.ResponseCode)
+	_, err := dw.writer.Write(dw.buffer[:dw.bytesBuffered])
+	if err != nil {
+		panic(err)
+	}
+	dw.writer.(http.Flusher).Flush()
+}
+
+// Header returns the header map that will be sent by WriteHeader.
+func (dw *DelayedWriter) Header() http.Header {
+	return dw.writer.Header()
+}
+
+// Write only writes the data to the buffer, not the output.
+func (dw *DelayedWriter) Write(p []byte) (int, error) {
+	if len(dw.buffer) == 0 {
+		dw.buffer = make([]byte, 32*1024)
```

Comment:
should this be hard coded?

Reply:
This is why I think this implementation is hacky: we are changing the semantics of QueueProxy.

The previous behavior is: `reverseproxy` will use a 32kB buffer to write output to the client (see [line 458](https://go.dev/src/net/http/httputil/reverseproxy.go) here), so if the actual response body size is 64kB, it will just write twice (and flush twice).

The current behavior is: with the delayed_writer, we are saving everything in memory (not like above where we save 32kB and flush, and then save another 32kB and flush...). Assuming we have a 32kB buffer here, if the actual response body size is 33kB, we will lose the extra 1kB. But if we increase the buffer size, we are increasing the memory consumption of QueueProxy.

Of course we should make the buffer size here configurable instead of hard coding it. But setting the buffer size to X KB basically means "Knative serving will only support user containers whose response body size is less than X KB". Do you think that makes sense?

Reply:
Just thinking out loud here (so there very well may be good reasons not to this), but if the goal is to track across components, would emitting a trace in the various components so that the timestamps could be compared, etc. provide what you're looking for? In other words, doing the measurement outside of queue-proxy rather than within?

Reply:
thats a good point @psschwei, we should probably try to achieve the original goal a different way, would something like a different metric being emitted by Prometheus to indicate the latency in the queue-proxy only help? or logged somehow

Reply:
Thanks for the suggestions! Let me explain why this is necessary for us. Our architecture is like this:
our own load balancer -> Istio -> Knative activator -> QueueProxy -> user container.

In the user container we are running customers' workloads and we don't care too much about the latency there. We mainly care about the "overhead latency", which is brought by our own load balancer, Istio, Knative activator and QueueProxy.

To measure the P99 value of the "overhead latency", we have to let QueueProxy report a header "user_container_latency", and in our own load balancer, we can do "overall_latency" minus "user_container_latency" and get the "overhead latency", and emit that "overhead latency" to Prometheus. That's how we can monitor the P99 of the overhead latency.

Of course we can insert one more proxy between QueueProxy and user container to report the "user_container_latency" header, but we just want to try to do that in QueueProxy first as that'll save us some time.

Reply:
yes, I understand what you are trying to find out, just wondering about if there is a way than can achieve that other than the header?

Reply:
Thanks! I'd love to know other ways that can achieve this without the header.

> emitting a trace in the various components so that the timestamps could be compared

Our custom load balancer is running on a different k8s cluster than Knative, so setting up tracing will be hard. Besides, tracing is for debugging individual requests, and we want to know the "overhead latency" of all requests here and monitor the P99.

> would something like a different metric being emitted by Prometheus to indicate the latency in the queue-proxy only help? or logged somehow

This will give us the P99 latency in the queue-proxy only. And we can do the same thing for the activator, Istio, and our custom load balancer to get the P99 latency within each component.

overhead latency = custom load balancer latency + Istio latency + activator latency + queue-proxy latency (+ network latency)
But P99 overhead latency != P99 custom load balancer latency + P99 Istio latency + P99 activator latency + P99 queue-proxy latency (+ network latency)

Basically P99 values are not additive. So I guess having this header is the only way?

Reply:
> The previous behavior is: reverseproxy will use a 32kB buffer to write output to the client (see line 458 here), so if the actual response body size is 64kB, it will just write twice (and flush twice).

I'm worried this hijacking even if the header is not required will slow down all the requests even by small bit. 

Maybe you can add your own queue-proxy instead of the default one? https://github.com/knative/serving/blob/39df601c36ebe589509fa28daa5b7abba525aed0/config/core/configmaps/deployment.yaml#L28-L31

Reply:
Having our own queue-proxy is a great idea! That means partially forking Knative (only forking the queue-proxy code), right? We'll lose the new features on queue-proxy added by others in the future. I'll talk to my teammates about it.

Reply:
Yes only the queue proxy, the property i linked above allows you to select a different image for the queue proxy container